File: 140. Word Break II/140_Original_DP_TLE.cs
public class Solution {
    public IList<string> WordBreak(string s, IList<string> wordDict) {        
        //DP approach
        var dp = new Dictionary<int, List<string>>();
        dp[0] = new List<string>{ "" };
        var hs = new HashSet<string>();
        foreach(var word in wordDict)
            hs.Add(word);
        var sb = new StringBuilder();
        for(var i = 0; i < s.Length; i++){
            if(dp.ContainsKey(i)){
                sb.Clear();
                for(var j = i; j < s.Length; j++){
                    sb.Append(s[j]);
                    var cur = sb.ToString();
                    if(hs.Contains(cur)){
                        if(!dp.ContainsKey(j + 1))
                            dp[j + 1] = new List<string>();
                        foreach(var prev in dp[i]){
                            dp[j + 1].Add(string.IsNullOrEmpty(prev) ? cur : prev + " " + cur);
                        }
                    }
                }
            }
        }
        return dp.ContainsKey(s.Length) ? dp[s.Length] : new List<string>();
    }
}

File: 140. Word Break II/140_Original_DP_TLE.java
class Solution {
    public List<String> wordBreak(String s, List<String> wordDict) {
        //DP approach
        Map<Integer, ArrayList<String>> dp = new HashMap<Integer, ArrayList<String>>();
        dp.put(0, new ArrayList<String>(Arrays.asList("")));
        Set hs = new HashSet<String>();
        for(String word : wordDict)
            hs.add(word);
        StringBuilder sb = new StringBuilder();
        for(Integer i = 0; i < s.length(); i++){
            if(dp.containsKey(i)){
                sb.setLength(0);
                for(Integer j = i; j < s.length(); j++){
                    sb.append(s.charAt(j));
                    var cur = sb.toString();
                    if(hs.contains(cur)){
                        if(!dp.containsKey(j + 1))
                            dp.put(j + 1, new ArrayList<String>());
                        for(String prev : dp.get(i)){
                            dp.get(j + 1).add(prev.isEmpty() ? cur : prev + " " + cur);
                        }
                    }
                }
            }
        }
        return dp.containsKey(s.length()) ? dp.get(s.length()) : new ArrayList<String>();
    }
}